src/Utilities/TaggedTuple.hpp

```diff
@@ -705,13 +705,61 @@ template <template <typename...> class List, typename... Tags>
 struct tagged_tuple_typelist_impl<List<Tags...>> {
   using type = TaggedTuple<Tags...>;
 };
+
+template <typename T>
+struct typelist_from_tagged_tuple_impl;
+
+template <typename... Tags>
+struct typelist_from_tagged_tuple_impl<TaggedTuple<Tags...>> {
+  using type = tmpl::list<Tags...>;
+};
 }  // namespace TaggedTuple_detail
 
 /// \ingroup UtilitiesGroup
 template <typename T>
 using tagged_tuple_from_typelist =
     typename TaggedTuple_detail::tagged_tuple_typelist_impl<T>::type;
 
+template <typename T>
+using typelist_from_tagged_tuple =
+    typename TaggedTuple_detail::typelist_from_tagged_tuple_impl<T>::type;
+
+namespace TaggedTuple_detail {
+
+template <typename T>
+struct reorder_tagged_tuple_impl;
+
+template <typename... Tags>
+struct reorder_tagged_tuple_impl<TaggedTuple<Tags...>> {
+  template <typename InputTaggedTuple>
+  static TaggedTuple<Tags...> apply(InputTaggedTuple&& input) noexcept {
+    static_assert(
+        cpp17::is_same_v<
+            tmpl::list_difference<tmpl::list<Tags...>,
+                                  typelist_from_tagged_tuple<InputTaggedTuple>>,
+            tmpl::list<>> and
+            cpp17::is_same_v<tmpl::list_difference<
+                                 typelist_from_tagged_tuple<InputTaggedTuple>,
+                                 tmpl::list<Tags...>>,
+                             tmpl::list<>>,
+        "The input and output TaggedTuples must be the same except"
+        "for ordering.");
+    return TaggedTuple<Tags...>(std::move(get<Tags>(input))...);
+  }
+};
+}  // namespace TaggedTuple_detail
+
+/// Given an input TaggedTuple, produce an output TaggedTuple
+/// with the tags in a different order.  All tags must be the same
+/// except for ordering.
+/// \example
+/// \snippet Test_TaggedTuple.cpp reorder_tagged_tuple_example
+template <typename Output, typename Input>
+Output reorder_tagged_tuple(Input&& input) noexcept {
```

Comment:
I would require the first template parameter to be a typelist of the output tags and name it `TagListInReturnOrder`.

You should also have two overloads of this function so you can type match early on (gives better compiler errors and in this case minimal code bloat)